sql/core/src/test/scala/org/apache/spark/sql/DataFrameSetOperationsSuite.scala

```diff
@@ -385,6 +385,90 @@ class DataFrameSetOperationsSuite extends QueryTest with SharedSparkSession {
     checkAnswer(unionDF, unionAllDF)
   }
 
+  test("SQL-style union using Dataset: remove unnecessary deduplicate in multiple unions") {
+    val unionDF = testData.union(testData).distinct().union(testData).distinct()
+      .union(testData).distinct().union(testData).distinct()
+
+    // Before optimizer, there are three 'union.deduplicate' operations should be combined.
+    assert(unionDF.queryExecution.analyzed.collect {
+      case j: Union if j.children.size == 4 => j
```

Comment:
nit: isn't it more intuitive to simply count the number of aggregates for `distinct`?

Reply:
IMO, it's more intuitional to count the number of Union, and the test cases related to Union above use this way to check.